### PR TITLE
Ensemble builder

### PR DESCRIPTION
**Pull request by franchuterivera**
Initial Commit with an empty file, mainly to test the check-api

### CODE REVIEW

File: autoPyTorch/ensemble/abstract_ensemble.py
Comment (ravinkohli) [inline autoPyTorch/ensemble/abstract_ensemble.py:27]: do you mean its between 0 and 1 in the case of regression?

File: autoPyTorch/ensemble/ensemble_builder.py
Comment (ravinkohli) [inline autoPyTorch/ensemble/ensemble_builder.py:630]: AutoPyTorch, :)

**Comment by ravinkohli** [comment]
Should we wait for the dataset class to be merged and then make changes to this PR before merging it? Otherwise, for me the changes are good.

**Comment by franchuterivera** [comment]
We need the changes in this PR for SMAC. It also fixes the backend

When building an ensemble we use smac output, so we do not care about the data manager. 

Nevertheless, I am reading the datamanager to make any time test performance, but is a 1 line change when the dataset/datamanager is ready

**Comment by ravinkohli** [comment]
> We need the changes in this PR for SMAC. It also fixes the backend
> 
> When building an ensemble we use SMAC output, so we do not care about the data manager.
> 
> Nevertheless, I am reading the data manager to make any time test performance, but is a 1 line change when the dataset/data manager is ready
Okay then, from my side it can be merged. Let's wait for Lucas to review it

File: autoPyTorch/data/abstract_data_manager.py
Comment (LMZimmer) [inline autoPyTorch/data/abstract_data_manager.py:7]: Maybe we could add some documentation here?

File: autoPyTorch/ensemble/ensemble_builder.py
Comment (LMZimmer) [inline autoPyTorch/ensemble/ensemble_builder.py:481]: Do we still need this?

File: autoPyTorch/ensemble/ensemble_builder.py
Comment (LMZimmer) [inline autoPyTorch/ensemble/ensemble_builder.py:582]: Could we add a docstring here?

File: autoPyTorch/ensemble/ensemble_builder.py
Comment (LMZimmer) [inline autoPyTorch/ensemble/ensemble_builder.py:656]: Could we maybe add a short docstring?

File: autoPyTorch/pipeline/base_pipeline.py
Comment (LMZimmer) [inline autoPyTorch/pipeline/base_pipeline.py:133]: Could we add an info that batch size is set automatically here?

**Review by LMZimmer — CHANGES_REQUESTED** [review]
Thanks for the work! Some points:

1. Looks like some files are geting through the .gitignore in `test/test_ensemble/.autoPyTorch`

2. We might want to provide a function to get the ensemble models/directories/pipelines since this is something many users wanted. Let's either provide that now or create an issue to keep this in mind

**Review by BryanLaura — APPROVED** [review]

**Review by BryanLaura — APPROVED** [review]

File: autoPyTorch/data/abstract_data_manager.py
Comment (franchuterivera) [reply autoPyTorch/data/abstract_data_manager.py:7]: This was a WA because we didn't have the dataset class! I am going to rebase and remove this which is no longer needed

File: autoPyTorch/ensemble/ensemble_builder.py
Comment (franchuterivera) [reply autoPyTorch/ensemble/ensemble_builder.py:481]: Yes and I have update the comment as to why:
```
  # already read prediction files
         # We read in back this object to give the ensemble the possibility to have memory
         # Every ensemble task is sent to dask as a function, that cannot take un-picklable
         # objects as attributes. For this reason, we dump to disk the stage of the past
         # ensemble iterations to kick-start the ensembling process

```

File: autoPyTorch/datasets/base_dataset.py
Comment (LMZimmer) [inline autoPyTorch/datasets/base_dataset.py:128]: Maybe you could add an `else:` with a warning/error when this is called after already creating a split?

File: autoPyTorch/datasets/base_dataset.py
Comment (LMZimmer) [inline autoPyTorch/datasets/base_dataset.py:166]: Maybe something like `split_id` is more descriptive (also since this occurs later in the pipeline as a config option)

File: autoPyTorch/datasets/base_dataset.py
Comment (LMZimmer) [inline autoPyTorch/datasets/base_dataset.py:195]: If you create the warning as mentioned above, this could have an `if self.splits is None:`

File: autoPyTorch/pipeline/components/training/data_loader/base_data_loader.py
Comment (LMZimmer) [inline autoPyTorch/pipeline/components/training/data_loader/base_data_loader.py:50]: Same as above, `split_id` might be more expressive

File: examples/example_ensemble_classification.py
Comment (LMZimmer) [inline examples/example_ensemble_classification.py:103]: `id` is a Python built-in, so I'd prefer another name (I think autosklearn also uses this)

**Review by LMZimmer — CHANGES_REQUESTED** [review]
Just very minor comments. Also, there still are a bunch of unnecessary files in `test/test_ensemble/data/.autoPyTorch` that should be removed (just delete them locally, commit and then maybe add `test/test_ensemble/data/.autoPyTorch` to the `.gitignore`)

File: autoPyTorch/datasets/base_dataset.py
Comment (franchuterivera) [reply autoPyTorch/datasets/base_dataset.py:128]: I have clarified why I do expect this function to be called multiple times via a comment:
```
      # Create just the split once
         # This is gonna be called multiple times, because the current dataset
         # is being used for multiple pipelines. That is, to be efficient with memory
         # we dump the dataset to memory and read it on a need basis. So this function
         # should be robust against multiple calls, and it does so by remembering the splits

```

File: autoPyTorch/datasets/base_dataset.py
Comment (franchuterivera) [reply autoPyTorch/datasets/base_dataset.py:195]: Ah, ok cool... processing this comment serially... so I will do that!

File: examples/example_ensemble_classification.py
Comment (franchuterivera) [reply examples/example_ensemble_classification.py:103]: I think this is not needed and I have removed it.

File: autoPyTorch/pipeline/components/training/data_loader/base_data_loader.py
Comment (franchuterivera) [reply autoPyTorch/pipeline/components/training/data_loader/base_data_loader.py:50]: Done!

**Pull request merged**
